File: CS/App_Code/ProductDataContext.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.Web;
using System.Linq;

[DataObject(true)]
public class ProductDataContext {
    private static List<Product> Products {
        get {
            if (HttpContext.Current.Session["Products"] == null) {
                List<Product> list = new List<Product>();
                list.Add(new Product(0, "Apples", 1, 10.0m));
                list.Add(new Product(1, "Gingerbread", 10, 15.0m));
                list.Add(new Product(2, "Yogurt", 12, 20.0m));
                HttpContext.Current.Session["Products"] = list;
                return list;
            }
            else
                return (List<Product>)HttpContext.Current.Session["Products"];
        }
        set {
            HttpContext.Current.Session["Products"] = value;
        }
    }
    [DataObjectMethod(DataObjectMethodType.Select, true)]
    public List<Product> GetProducts() {
        return Products;
    }

    [DataObjectMethod(DataObjectMethodType.Insert, true)]
    public static int InsertProduct(Product product) {
        List<Product> products = Products;
        product.ProductID = Products.Count == 0 ? 0 : Products.Max(i => i.ProductID) + 1;
        products.Add(product);
        return product.ProductID;
    }

    [DataObjectMethod(DataObjectMethodType.Update, true)]
    public static void UpdateProduct(Product product) {
        List<Product> products = Products;
        var productToUpdate = products.Find(p => p.ProductID == product.ProductID);
        productToUpdate.ProductName = product.ProductName;
        productToUpdate.UnitPrice = product.UnitPrice;
        productToUpdate.UnitsOnOrder = product.UnitsOnOrder;
    }

    [DataObjectMethod(DataObjectMethodType.Delete, true)]
    public static void DeleteProduct(Product product) {
        List<Product> products = Products;
        products.Remove(product);
    }
}

File: VB/App_Code/ProductDataContext.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Web
Imports System.Linq

<DataObject(True)> _
Public Class ProductDataContext
    Private Shared Property Products() As List(Of Product)
        Get
            If HttpContext.Current.Session("Products") Is Nothing Then
                Dim list As New List(Of Product)()
                list.Add(New Product(0, "Apples", 1, 10.0D))
                list.Add(New Product(1, "Gingerbread", 10, 15.0D))
                list.Add(New Product(2, "Yogurt", 12, 20.0D))
                HttpContext.Current.Session("Products") = list
                Return list
            Else
                Return DirectCast(HttpContext.Current.Session("Products"), List(Of Product))
            End If
        End Get
        Set(ByVal value As List(Of Product))
            HttpContext.Current.Session("Products") = value
        End Set
    End Property
    <DataObjectMethod(DataObjectMethodType.Select, True)> _
    Public Function GetProducts() As List(Of Product)
        Return Products
    End Function

    <DataObjectMethod(DataObjectMethodType.Insert, True)> _
    Public Shared Function InsertProduct(ByVal product As Product) As Integer

        Dim products_Renamed As List(Of Product) = Products
        product.ProductID = If(Products.Count = 0, 0, Products.Max(Function(i) i.ProductID) + 1)
        products_Renamed.Add(product)
        Return product.ProductID
    End Function

    <DataObjectMethod(DataObjectMethodType.Update, True)> _
    Public Shared Sub UpdateProduct(ByVal product As Product)

        Dim products_Renamed As List(Of Product) = Products
        Dim productToUpdate = products_Renamed.Find(Function(p) p.ProductID = product.ProductID)
        productToUpdate.ProductName = product.ProductName
        productToUpdate.UnitPrice = product.UnitPrice
        productToUpdate.UnitsOnOrder = product.UnitsOnOrder
    End Sub

    <DataObjectMethod(DataObjectMethodType.Delete, True)> _
    Public Shared Sub DeleteProduct(ByVal product As Product)

        Dim products_Renamed As List(Of Product) = Products
        products_Renamed.Remove(product)
    End Sub
End Class